File: frontends/Python/exasim/Postprocessing/readsurfacequantities.py
import os
import numpy


def _readbou(filename):
    """Read one outbou*_np<r>.bin file: (3-double header, flat float64 payload)."""
    tm = numpy.fromfile(filename, dtype=numpy.float64)
    if tm.size < 3:
        raise ValueError("readsurfacequantities: %s has no header" % filename)
    return numpy.int_(numpy.round(tm[0:3])), tm[3:]


def _splitblocks(data, blocks, npts, ncomp):
    """Split a per-block concatenation into {ib: [npts, nf_b, ncomp] arrays} (in file order)."""
    out = []
    k = 0
    for ib, nf in blocks:
        n = npts * nf * ncomp
        out.append((ib, numpy.reshape(data[k:k + n], (npts, nf, ncomp), order='F')))
        k += n
    return out, k


def readsurfacequantities(fileprefix, nranks=1, saveSolBouLoc=None):
    """Read the SurfaceQuantities written on the ibs boundaries (pde['saveSolBouFreq'] > 0).

    Parameters
    ----------
    fileprefix : str
        Output path prefix, e.g. os.path.join(dataoutpath, "out"); files are
        <fileprefix>bouinfo_np<r>.bin, <fileprefix>bousurf_np<r>.bin, ...
    nranks : int
        Number of MPI ranks (files np0 .. np<nranks-1>).
    saveSolBouLoc : None, 0 or 1
        Evaluation location used by the run. None (default) infers it: 1 when a
        bousurfgeo file exists (or the point count differs from npf), else 0. Pass it
        explicitly if the output directory may hold stale files from another run.

    Returns
    -------
    dict keyed by boundary id ib, each entry a dict with
        'values' : [np, nf, nsurfq, nsteps] surface quantities (np = npf face nodes when
                   saveSolBouLoc = 0, ngf face Gauss points when saveSolBouLoc = 1)
        'x'      : [np, nf, ncx] coordinates of those points
        'n'      : [np, nf, nd]  unit normals at those points
        'dA'     : [np, nf] quadrature weight * face jacobian (saveSolBouLoc = 1 only, else None),
                   so (values[:, :, i, t] * dA).sum() integrates quantity i over the boundary
        'saveSolBouLoc' : 0 or 1
    Faces are concatenated across ranks (rank order) and face blocks with the same ib.
    """
    parts = {}
    for r in range(nranks):
        hdr, info = _readbou(fileprefix + "bouinfo_np" + str(r) + ".bin")
        nblk = hdr[0]
        blocks = [(int(round(info[2 * j])), int(round(info[2 * j + 1]))) for j in range(nblk)]
        if nblk == 0:
            continue  # this rank owns no face on the ibs boundaries
        nfbou = sum(nf for _, nf in blocks)

        hs, sdata = _readbou(fileprefix + "bousurf_np" + str(r) + ".bin")
        npts, nfs, nsurfq = hs
        if nfs != nfbou:
            raise ValueError("readsurfacequantities: bousurf/bouinfo face counts differ on rank %d" % r)
        reclen = npts * nfbou * nsurfq
        nsteps = sdata.size // reclen if reclen > 0 else 0
        recs = [_splitblocks(sdata[t * reclen:(t + 1) * reclen], blocks, npts, nsurfq)[0]
                for t in range(nsteps)]

        hx, xdata = _readbou(fileprefix + "bouxdg_np" + str(r) + ".bin")
        hn, ndata = _readbou(fileprefix + "boundg_np" + str(r) + ".bin")
        npf, ncx, nd = hx[0], hx[2], hn[2]
        geofile = fileprefix + "bousurfgeo_np" + str(r) + ".bin"
        if saveSolBouLoc is None:
            loc = 1 if (npts != npf or os.path.isfile(geofile)) else 0
        else:
            loc = int(saveSolBouLoc)
        if loc == 1:
            # saveSolBouLoc = 1: Gauss-point geometry, per block [xg | nlg | dA].
            hg, gdata = _readbou(geofile)
            ngf, ncx, nd = hg[0], hg[2] - 1 - hn[2], hn[2]
            if ngf != npts:
                raise ValueError("readsurfacequantities: bousurfgeo/bousurf point counts differ on rank %d" % r)
            geo = []
            k = 0
            for ib, nf in blocks:
                n = ngf * nf
                xg = numpy.reshape(gdata[k:k + n * ncx], (ngf, nf, ncx), order='F'); k += n * ncx
                ng = numpy.reshape(gdata[k:k + n * nd], (ngf, nf, nd), order='F'); k += n * nd
                dA = numpy.reshape(gdata[k:k + n], (ngf, nf), order='F'); k += n
                geo.append((ib, xg, ng, dA))
        else:
            xb = _splitblocks(xdata, blocks, npf, ncx)[0]
            nb = _splitblocks(ndata, blocks, npf, nd)[0]
            geo = [(xb[j][0], xb[j][1], nb[j][1], None) for j in range(len(blocks))]

        for j, (ib, nf) in enumerate(blocks):
            p = parts.setdefault(ib, {'values': [], 'x': [], 'n': [], 'dA': [], 'saveSolBouLoc': loc})
            s = numpy.zeros((npts, nf, nsurfq, nsteps))
            for t in range(nsteps):
                s[:, :, :, t] = recs[t][j][1]
            p['values'].append(s)
            p['x'].append(geo[j][1])
            p['n'].append(geo[j][2])
            p['dA'].append(geo[j][3])

    out = {}
    for ib, p in parts.items():
        out[ib] = {
            'values': numpy.concatenate(p['values'], axis=1),
            'x': numpy.concatenate(p['x'], axis=1),
            'n': numpy.concatenate(p['n'], axis=1),
            'dA': numpy.concatenate(p['dA'], axis=1) if p['saveSolBouLoc'] == 1 else None,
            'saveSolBouLoc': p['saveSolBouLoc'],
        }
    return out

File: frontends/Julia/Exasim/src/Postprocessing/readsurfacequantities.jl
"""
    readsurfacequantities(prefix, nranks, saveSolBouLoc=nothing) -> Dict{Int,Dict{Symbol,Any}}

Read the pointwise surface quantities written by the backend when the model
defines `surfacequantities` and `pde.saveSolBouFreq > 0` (boundaries `pde.ibs`).

`prefix` is the output prefix, e.g. `joinpath(pde.datapath, "dataout", "out")`;
the files `prefix * "bouinfo_np<r>.bin"`, `prefix * "bousurf_np<r>.bin"`, ...
are read for ranks r = 0..nranks-1 (ranks without an info file are skipped).
`saveSolBouLoc` (0 face nodes, 1 face Gauss points) is the run's evaluation
location; `nothing` infers 1 when `outbousurfgeo_np0.bin` exists. Pass it
explicitly if the output directory may hold stale files from another run.

Returns a Dict keyed by boundary index `ib`, each entry a Dict with
- `:values` `[np, nf, nsurfq, nsteps]` (np = npf for saveSolBouLoc = 0, ngf for 1)
- `:x`      `[np, nf, ncx]` coordinates of those points
- `:n`      `[np, nf, nd]`  unit normals at those points
- `:dA`     `[np, nf]` Gauss weight * face jacobian (saveSolBouLoc = 1 only, else
            `nothing`), so `sum(values[:, :, i, t] .* dA)` integrates quantity i
- `:saveSolBouLoc` 0 or 1
Faces are concatenated over ranks and blocks with the same `ib`.
"""
function readsurfacequantities(prefix::AbstractString, nranks::Integer, saveSolBouLoc=nothing)
    rd(fn) = open(fn, "r") do io
        read!(io, Vector{Float64}(undef, filesize(fn) ÷ 8))
    end
    loc = saveSolBouLoc === nothing ? Int(isfile(prefix * "bousurfgeo_np0.bin")) : Int(saveSolBouLoc)
    loc in (0, 1) || error("readsurfacequantities: saveSolBouLoc must be 0 or 1")
    out = Dict{Int,Dict{Symbol,Any}}()
    catdim(a, b, d) = a === nothing ? b : cat(a, b; dims=d)
    function add!(ib, key, arr, d)
        e = get!(out, ib, Dict{Symbol,Any}(:dA => nothing, :saveSolBouLoc => loc))
        e[key] = catdim(get(e, key, nothing), arr, d)
    end
    for r in 0:nranks-1
        finfo = prefix * "bouinfo_np$(r).bin"
        isfile(finfo) || continue
        info = rd(finfo)
        nblk = Int(info[1])
        blocks = [(Int(info[3+2k+1]), Int(info[3+2k+2])) for k in 0:nblk-1]
        nftot = sum(b[2] for b in blocks; init=0)
        nftot > 0 || continue   # this rank owns no face on the ibs boundaries

        fsurf = prefix * "bousurf_np$(r).bin"
        a = rd(fsurf)
        np, nfb, nsq = Int(a[1]), Int(a[2]), Int(a[3])
        nfb == nftot || error("readsurfacequantities: $fsurf has nfbou=$nfb, bouinfo says $nftot")
        reclen = np * nfb * nsq
        data = a[4:end]
        nsteps = reclen > 0 ? length(data) ÷ reclen : 0
        data = reshape(data[1:reclen*nsteps], reclen, nsteps)
        off = 0
        for (ib, nf) in blocks
            n = np * nf * nsq
            add!(ib, :values, reshape(data[off+1:off+n, :], np, nf, nsq, nsteps), 2)
            off += n
        end

        # outboundg's header gives nd; outbouxdg's gives ncx.
        hx, hn = rd(prefix * "bouxdg_np$(r).bin"), rd(prefix * "boundg_np$(r).bin")
        ncx, nd = Int(hx[3]), Int(hn[3])
        if loc == 0
            # face nodes: the geometry is exactly outbouxdg / outboundg
            for (a, key, nc) in ((hx, :x, ncx), (hn, :n, nd))
                npf = Int(a[1])
                npf == np || error("readsurfacequantities: rank $r has $np points per face but npf=$npf; pass saveSolBouLoc=1?")
                off = 3
                for (ib, nf) in blocks
                    n = npf * nf * nc
                    add!(ib, key, reshape(a[off+1:off+n], npf, nf, nc), 2)
                    off += n
                end
            end
        else
            # face Gauss points: per block [x | n | dA] from outbousurfgeo
            a = rd(prefix * "bousurfgeo_np$(r).bin")
            ngf = Int(a[1])
            ngf == np || error("readsurfacequantities: bousurfgeo/bousurf point counts differ on rank $r")
            off = 3
            for (ib, nf) in blocks
                m = ngf * nf
                add!(ib, :x, reshape(a[off+1:off+m*ncx], ngf, nf, ncx), 2); off += m*ncx
                add!(ib, :n, reshape(a[off+1:off+m*nd], ngf, nf, nd), 2);   off += m*nd
                add!(ib, :dA, reshape(a[off+1:off+m], ngf, nf), 2);          off += m
            end
        end
    end
    return out
end
